roc.go

```diff
@@ -0,0 +1,110 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package stat
+
+import "sort"
+
+// ROC returns paired false positive rate (FPR) and true positive rate
+// (TPR) values corresponding to n cutoffs spanning the relative
+// (or receiver) operator characteristic (ROC) curve obtained when y is
+// treated as a binary classifier for classes with weights.
+//
+// Cutoffs are equally spaced from eps less than the minimum value of y
+// to the maximum value of y, including both endpoints -- this means the
+// resulting ROC curve will always begin at (0,0) and end at (1,1).
+//
+// The input y must be sorted, and SortWeightedLabeled can be used in
+// order to sort y together with classes and weights.
+//
+// For a given cutoff value, observations corresponding to entries in y
+// greater than the cutoff value are classified as false, while those
+// below (or equal to) the cutoff value are classified as true. These
+// assigned class labels are compared with the true values in the classes
+// slice and used to calculate the FPR and TPR.
+//
+// If weights is nil, all weights are treated as 1.
+//
+// n of zero results in all possible cutoffs being calculated -- this will
+// result in fpr and tpr having length len(y)+1 if y has no duplicated
+// values and one greater than the number of unique values in y in
+// general. n >= 2 will result in fpr and tpr having length n.
+// n of one or negative values of n are not supported.
+//
+// More details about ROC curves are available at
+// https://en.wikipedia.org/wiki/Receiver_operating_characteristic
+func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []float64) {
+	if len(y) != len(classes) {
+		panic("stat: slice length mismatch")
+	}
+	if weights != nil && len(y) != len(weights) {
+		panic("stat: slice length mismatch")
+	}
+	if !sort.Float64sAreSorted(y) {
+		panic("stat: input must be sorted")
+	}
+
+	var incWidth, tol float64
+	if n != 0 {
+		if n < 2 {
+			panic("stat: cannot calculate fewer than 2 points on a ROC curve")
+		}
+		incWidth = (y[len(y)-1] - y[0]) / float64(n-1)
+		if incWidth <= 0 {
+			panic("stat: input constant")
+		}
+		tol = y[0] + incWidth
+		tpr = make([]float64, n)
+		fpr = make([]float64, n)
+	} else {
+		tpr = make([]float64, len(y)+1)
+		fpr = make([]float64, len(y)+1)
+	}
+	var bin int = 1 // the initial bin is known to have 0 fpr and 0 tpr
+	var nPos, nNeg, posWeight, negWeight float64
+	for i, u := range classes {
+
+		posWeight, negWeight = 0, 1
+		if weights != nil {
+			negWeight = weights[i]
+		}
+		if u {
+			posWeight, negWeight = negWeight, posWeight
+		}
+		nPos += posWeight
+		nNeg += negWeight
+		tpr[bin] += posWeight
+		fpr[bin] += negWeight
+
```

Comment:
A more elaborate comment would be clearer up here than the half comments in the if.

```` Assess if the bin needs to be updated. If n is zero, the bin is always updated unless the consecutive y values are equal. Otherwise, the bin must be evaluated if the y value is sufficiently large````